ofxconnect/ofxconnect.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.4.
.TH LIBOFX "1" "April 2011" "libofx 0.9.4" "User Commands"
.SH NAME
libofx \- manual page for libofx 0.9.4
.SH SYNOPSIS
.B libofx
[\fIOPTIONS\fR]... [\fIFILES\fR]...
.SH DESCRIPTION
libofx 0.9.4
.PP
prints to stdout the created OFX file based on the options you pass it.
currently it will only create a statement request file.  you can POST this to
an OFX server to request a statement from that financial institution for that
account.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.TP
\fB\-\-fipid\fR=\fISTRING\fR
FI partner identifier (looks up fid, org & url from
partner server)
.TP
\fB\-\-fid\fR=\fISTRING\fR
FI identifier
.TP
\fB\-\-org\fR=\fISTRING\fR
FI org tag
.TP
\fB\-\-bank\fR=\fISTRING\fR
IBAN bank identifier
.TP
\fB\-\-broker\fR=\fISTRING\fR
Broker identifier
.TP
\fB\-\-user\fR=\fISTRING\fR
User name
.TP
\fB\-\-pass\fR=\fISTRING\fR
Password
.TP
\fB\-\-acct\fR=\fISTRING\fR
Account ID
.TP
\fB\-\-type\fR=\fIINT\fR
Account Type 1=checking 2=invest 3=ccard
.TP
\fB\-\-past\fR=\fILONG\fR
How far back to look from today (in days)
.TP
\fB\-\-url\fR=\fISTRING\fR
Url to POST the data to (otherwise goes to stdout)
.TP
\fB\-\-trid\fR=\fIINT\fR
Transaction id
.IP
Group: command
.TP
\fB\-s\fR, \fB\-\-statement\-req\fR
Request for a statement
.TP
\fB\-a\fR, \fB\-\-accountinfo\-req\fR
Request for a list of accounts
.TP
\fB\-p\fR, \fB\-\-payment\-req\fR
Request to make a payment
.TP
\fB\-i\fR, \fB\-\-paymentinquiry\-req\fR
Request to inquire about the status of a payment
.TP
\fB\-b\fR, \fB\-\-bank\-list\fR
List all known banks
.TP
\fB\-f\fR, \fB\-\-bank\-fipid\fR
List all fipids for a given bank
.TP
\fB\-v\fR, \fB\-\-bank\-services\fR
List supported services for a given fipid
.TP
\fB\-\-allsupport\fR
List all banks which support online banking
.SH "SEE ALSO"
The full documentation for
.B libofx
is maintained as a Texinfo manual.  If the
.B info
and
.B libofx
programs are properly installed at your site, the command
.IP
.B info libofx
.PP
should give you access to the complete manual.


debian/changelog
libofx (1:0.9.4-0.1) unstable; urgency=low

  * Non-maintainer upload.
  * New upstream release (Closes: #549161, #551057)
  * Remove obsolete debian patches
  * Change libxml dependency to libxml++2.6-dev (Closes: #467364)
  * Build using libcurl4-gnutls-dev (Closes: #487255)
  * Add new help2man dependency
  * Add ${misc:Depends} entries to binary dependencies
  * Add debian/watch file
  * Back up manpages prior to build, as the upstream makefiles regenerate
    them, adding noise to the diffs
  * Don't ignore make clean failures
  * Update debhelper compatibility level to 5
  * Add versioned dependency on debhelper
  * Update debian/copyright to specifically reference GPL-2 rather than GPL in
    common-licenses
  * Update to Standards-Version: 3.9.2.0
    * Added Homepage: field to control file
  * Unversion ofx's conflict with libofx2; it's been long enough that only
    libofx4 should be in the wild. 
  * Update libofx-dev description 
  * Switch to 3.0 (quilt) source format
  * Use dh_installman to install manual pages 
  * Add doc-base registration for libofx-dev HTML reference docs

 -- Bryan Donlan <bdonlan@gmail.com>  Tue, 03 May 2011 21:33:29 -0400

libofx (1:0.9.0-3) unstable; urgency=low

  * lib/ofx_preproc.c (ofx_proc_file): Make iconv buffer twice as big in
    case input files include non-ASCII UTF-8 characters.  Patch thanks to
    Jerome Vouillon (jerome.Vouillon@pps.jussieu.fr).  (Closes: 493597)
  
  * Acknowledge NMU from Andreas Barth. (Thanks Andreas!) (Closes: #417361)
  * Acknowledge NMU from Chris Lamb. (Thanks Chris!) (Closes: #474860)

 -- Thomas Bushnell, BSG <tb@debian.org>  Mon, 04 Aug 2008 23:16:30 -0700

libofx (1:0.9.0-2.2) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with GCC 4.3 (Closes: #474860)
  * debian/control:
    - Bump Standards-Version to 3.7.3.
    - Use ${binary:Version} instead of ${Source-Version}.
  * Call distclean in debian/rules clean target.

 -- Chris Lamb <chris@chris-lamb.co.uk>  Sat, 12 Apr 2008 14:44:55 +0100

libofx (1:0.9.0-2.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with gcc-4.3. Thanks to Martin Michlmayr. Closes: #417361

 -- Andreas Barth <aba@not.so.argh.org>  Sun, 16 Mar 2008 22:48:14 +0000

libofx (1:0.9.0-2) unstable; urgency=low

  * debian/control (libofx4): Don't Conflict, Provide, or Replace
    the libofx2 packages.  That job is for the old libofx3 only.
    (Closes: #463854)

 -- Thomas Bushnell, BSG <tb@debian.org>  Mon, 04 Feb 2008 19:35:39 -0500

libofx (1:0.9.0-1) unstable; urgency=low

  * New upstream release.
  * Regenerate autotools using Debian libtool by:
    libtoolize --copy --force; aclocal-1.9; autoconf.
  * debian/control (libofx4): Renamed package from libofx3.
    (libofx-dev): Depend on libofx4 instead of libofx3.
  * debian/rules (PACKNAME): Update to libofx4.
    (version): Update to 0.9.0.
  * libofx4.docs: Renamed from libofx3.docs.
  
  * debian/rules (major): Delete unused variable, and commented idle
    code that mentioned it.
  
  * debian/copyright: Add proper authorship and copyright information, and
    the license information for the OFX dtd's.  (Closes: #460407)
  
  * debian/rules (binary-arch): Don't put /usr/include in the $(PACKNAME)
    package.
  
 -- Thomas Bushnell, BSG <tb@debian.org>  Sun, 30 Dec 2007 21:24:51 -0800

libofx (1:0.8.2-3) unstable; urgency=low

  * debian/control (Standards-Version): Update to 3.7.2.

 -- Thomas Bushnell, BSG <tb@debian.org>  Fri, 20 Oct 2006 13:12:33 -0700

libofx (1:0.8.2-2) unstable; urgency=low

  * Acknowledge NMU.  Thanks Matthias!  (Closes: #385821, #385667,
    #385880, #386068)

 -- Thomas Bushnell, BSG <tb@debian.org>  Thu, 28 Sep 2006 15:21:16 -0700

libofx (1:0.8.2-1.1) unstable; urgency=low

  * NMU.
  * Drop build dependency on g++-4.0. Closes: #385821.
  * New library package libofx3. Closes: #385667, #385880.

 -- Matthias Klose <doko@debian.org>  Tue,  5 Sep 2006 02:29:16 +0200

libofx (1:0.8.2-1) unstable; urgency=low

  * New upstream release.  (Closes: #287388)
  * debian/control (Build-Depends): Add libxml++1.0-dev.
  * Regenerate autotools using Debian libtool by:
    libtoolize --copy --force; aclocal-1.9; autoconf.
  * debian/rules: Put pkgconfig files into the -dev package.
  
 -- Thomas Bushnell, BSG <tb@debian.org>  Tue, 29 Aug 2006 12:52:48 -0700

libofx (1:0.8.0-12) unstable; urgency=low

  * debian/rules (MAKEVARS): Include $(version) in the name of the
    LIBOFX_DTD_DIR, so that we can put those files in the shared library
    package and still have more than one version of the shared library
    installed.
    (binary-arch): Put the dtd files into the shared library package.
  * debian/control (ofx): Don't mention the runtime support files here (or
    anywhere else, actually).
    
 -- Thomas Bushnell, BSG <tb@debian.org>  Mon, 29 May 2006 11:23:51 -0700

libofx (1:0.8.0-11) unstable; urgency=low

  * debian/control (libofx2c2a): No longer depend on ofx. This was only
    there because the old library package had the binaries which are now
    in the separate package.  These programs were not important enough for
    this to be done, and circular dependencies are bad.  Instead, use
    "Suggests". (Closes: #358383, #345672) 
    (libofx-dev): Conflict with libofx1, which contained some development
    files.
  
  * debian/ofx.docs: New file.
  * debian/libofx2c2a.docs: Renamed from libofx2.docs.
  * debian/libofx2.substvars, debian/libofx2.postinst.debhelper,
    debian/libofx2.postrm.debhelper: Remove cruft.
    (Closes: #361145)

 -- Thomas Bushnell, BSG <tb@debian.org>  Fri, 28 Apr 2006 18:41:04 -0700

libofx (1:0.8.0-10) unstable; urgency=low

  * debian/rules (PACKNAME, MAKEVARS): New variables, to reduce
    redundancy.  $(PACKNAME) replaces libofx2c2a.  $(MAKEVARS) is prefixed
    on all invocations of make.  Specify LIBOFX_DTD_DIR on make
    invocations.
    (install): No longer move files into correct location, since

ofxconnect/ofxconnect.cpp
/***************************************************************************
         ofx_connect.cpp 
                             -------------------
    copyright            : (C) 2005 by Ace Jones
    email                : acejones@users.sourceforge.net
***************************************************************************/
/**@file
 * \brief Code for ofxconnect utility.  C++ example code
 *
 * the purpose of the ofxconnect utility is to server as example code for
 * ALL functions of libOFX that have to do with creating OFX files.
 *
 * ofxconnect prints to stdout the created OFX file based on the options
 * you pass it
 *
 * currently it will only create the statement request file.  you can POST
 * this to an OFX server to request a statement from that financial
 * institution for that account.
 *
 * In the hopefully-not-to-distant future, ofxconnect will also make the
 * connection to the OFX server, post the data, and call ofxdump itself.
 */
/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>
#include "libofx.h"
#include <config.h>		/* Include config constants, e.g., VERSION TF */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <cstring>
#include <cstdlib>
#include <string.h>
#ifdef HAVE_LIBCURL
#include <curl/curl.h>
#endif

#include "cmdline.h" /* Gengetopt generated parser */

#include "nodeparser.h"
#include "ofxpartner.h"

using namespace std;

#ifdef HAVE_LIBCURL
bool post(const char* request, const char* url, const char* filename)
{
  CURL *curl = curl_easy_init();
  if(! curl)
    return false;

  unlink("tmpout");  
  FILE* file = fopen(filename,"wb");
  if (! file )
  {
    curl_easy_cleanup(curl);
    return false;
  }
    
  curl_easy_setopt(curl, CURLOPT_URL, url);
  curl_easy_setopt(curl, CURLOPT_POSTFIELDS, request);

  struct curl_slist *headerlist=NULL;
  headerlist=curl_slist_append(headerlist, "Content-type: application/x-ofx");
  headerlist=curl_slist_append(headerlist, "Accept: */*, application/x-ofx");    
  
  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headerlist);
  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fwrite);
  curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)file);
    
  CURLcode res = curl_easy_perform(curl);

  curl_easy_cleanup(curl);
  curl_slist_free_all (headerlist);
  
  fclose(file);
  
  return true;
}
#else
bool post(const char*, const char*, const char*)
{
  cerr << "ERROR: libox must be configured with libcurl to post this request directly" << endl;
  return false;
}
#endif

ostream& operator<<(ostream& os,const vector<string>& strvect)
{
  for( vector<string>::const_iterator it=strvect.begin(); it!=strvect.end(); ++it)
  {
    os << (*it) << endl;
  }
  return os;
}

int main (int argc, char *argv[])
{
  gengetopt_args_info args_info;
  
  if (cmdline_parser (argc, argv, &args_info) != 0)
    exit(1) ;

  if ( argc == 1 )
  {
    cmdline_parser_print_help();
    exit(1);
  }

  if ( args_info.statement_req_given || args_info.accountinfo_req_given )
  {
  if ( (args_info.inputs_num > 0) )
  {
    cout << "file " << args_info.inputs[0] << endl;
  }
  else
  {
    cerr << "ERROR: You must specify an output file" << endl;
  }
  }
  else if ( args_info.bank_fipid_given || args_info.bank_services_given )
  {
  if ( (args_info.inputs_num > 0) )
  {
    cout << "bank " << args_info.inputs[0] << endl;
  }
  else
  {
    cerr << "ERROR: You must specify an bank" << endl;
  }
  }
 
  OfxFiLogin fi;
  memset(&fi,0,sizeof(OfxFiLogin));
  bool ok = true;
  string url;
 
  if ( args_info.statement_req_given || args_info.accountinfo_req_given || args_info.payment_req_given || args_info.paymentinquiry_req_given )
  {
  // Get the FI Login information
  // 
 
  if ( args_info.fipid_given )
  {
    cerr << "fipid " <<  args_info.fipid_arg << endl;  
    cerr << "contacting partner server..." << endl;
    OfxFiServiceInfo svcinfo = OfxPartner::ServiceInfo(args_info.fipid_arg);
    cout << "fid " << svcinfo.fid << endl;
    strncpy(fi.fid,svcinfo.fid,OFX_FID_LENGTH-1);
    cout << "org " << svcinfo.org << endl;
    strncpy(fi.org,svcinfo.org,OFX_ORG_LENGTH-1);
    cout << "url " << svcinfo.url << endl;
    url = svcinfo.url;
  }
  if ( args_info.fid_given )
  {
    cerr << "fid " <<  args_info.fid_arg << endl;  
    strncpy(fi.fid,args_info.fid_arg,OFX_FID_LENGTH-1);
  }
  else if ( ! args_info.fipid_given )
  {
    cerr << "ERROR: --fid is required" << endl;
    ok = false;
  }
  
  if ( args_info.org_given )
  {
    cerr << "org " << args_info.org_arg << endl;  
    strncpy(fi.org,args_info.org_arg,OFX_ORG_LENGTH-1);
  }
  else if ( ! args_info.fipid_given )
  {
    cerr << "ERROR: --org is required" << endl;
    ok = false;
  }

  if ( args_info.user_given )
  {
    cerr << "user " << args_info.user_arg << endl;  
    strncpy(fi.userid,args_info.user_arg,OFX_USERID_LENGTH-1);
  }
  else
  {
    cerr << "ERROR: --user is required" << endl;
    ok = false;
  }
  
  if ( args_info.pass_given )
  {
    cerr << "pass " << args_info.pass_arg << endl;  
    strncpy(fi.userpass,args_info.pass_arg,OFX_USERPASS_LENGTH-1);
  }
  else
  {
    cerr << "ERROR: --pass is required" << endl;
    ok = false;
  }
  
  if ( args_info.url_given )
    url = args_info.url_arg;
  }
  
  if ( args_info.statement_req_given )
  {
    cerr << "Statement request" << endl;
    
    OfxAccountData account;
    memset(&account,0,sizeof(OfxAccountData));
    
    if ( args_info.bank_given )
    {
      cerr << "bank " << args_info.bank_arg << endl;  
      strncpy(account.bank_id,args_info.bank_arg,OFX_BANKID_LENGTH-1);
    }
    else    
    {
      if ( args_info.type_given && args_info.type_arg == 1 )
      {
        cerr << "ERROR: --bank is required for a bank request" << endl;
        ok = false;
      }
    }
    
    if ( args_info.broker_given )
    {
      cerr << "broker " << args_info.broker_arg << endl;  
      strncpy(account.broker_id,args_info.broker_arg,OFX_BROKERID_LENGTH-1);
    }
    else
    {
      if ( args_info.type_given && args_info.type_arg == 2 )
      {
        cerr << "ERROR: --broker is required for an investment statement request" << endl;
        ok = false;
      }
    }
    
    if ( args_info.acct_given )
    {
      cerr << "acct " << args_info.acct_arg << endl;  
      strncpy(account.account_number,args_info.acct_arg,OFX_ACCTID_LENGTH-1);
    }
    else
    {
      cerr << "ERROR: --acct is required for a statement request" << endl;
      ok = false;
    }
    
    if ( args_info.type_given )
    {
      cerr << "type " << args_info.type_arg << endl;
      switch (args_info.type_arg) {
      case 1: account.account_type = account.OFX_CHECKING;
	break;
      case 2: account.account_type = account.OFX_INVESTMENT;
	break;
      case 3: account.account_type = account.OFX_CREDITCARD ;
	break;
      default:
	cerr << "ERROR: --type is not valid.  Must be between 1 and 3" << endl;
	ok = false;
      }
    }
    else
    {
      cerr << "ERROR: --type is required for a statement request" << endl;
      ok = false;
    }
    
    if ( args_info.past_given )
    {
      cerr << "past " << args_info.past_arg << endl;  
    }
    else
    {
      cerr << "ERROR: --past is required for a statement request" << endl;
      ok = false;
    }
    
    if ( ok )
    {
      char* request = libofx_request_statement( &fi, &account, time(NULL) - args_info.past_arg * 86400L );
    
      if ( url.length() ) 
        post(request,url.c_str(),args_info.inputs[0]);
      else
        cout << request;
      
      free(request);
    }
  }

  if ( args_info.paymentinquiry_req_given )
  {
    char tridstr[33];
    memset(tridstr,0,33);

    bool ok = true;

    if ( args_info.trid_given )
    {
      cerr << "trid " << args_info.trid_arg << endl;  
      snprintf(tridstr,32,"%i",args_info.trid_arg);
    }
    else
    {
      cerr << "ERROR: --trid is required for a payment inquiry request" << endl;
      ok = false;
    }
 
    if ( ok )
    {
      char* request = libofx_request_payment_status( &fi, tridstr );
 
      filebuf fb;
      fb.open ("query",ios::out);
      ostream os(&fb);
      os << request;
      fb.close();
      
      if ( url.length() ) 
        post(request,url.c_str(),args_info.inputs[0]);
      else
        cout << request;
    
      free(request);
    }
  }
  
  if ( args_info.payment_req_given )
  {
    OfxAccountData account;
    memset(&account,0,sizeof(OfxAccountData));
    OfxPayee payee;
    memset(&payee,0,sizeof(OfxPayee));
    OfxPayment payment;
    memset(&payment,0,sizeof(OfxPayment));

    strcpy(payee.name,"MARTIN PREUSS");
    strcpy(payee.address1,"1 LAUREL ST");
    strcpy(payee.city,"SAN CARLOS");
    strcpy(payee.state,"CA");
    strcpy(payee.postalcode,"94070");
    strcpy(payee.phone,"866-555-1212");
        
    strcpy(payment.amount,"200.00");
    strcpy(payment.account,"1234");
    strcpy(payment.datedue,"20060301");
    strcpy(payment.memo,"This is a test");

    bool ok = true;

    if ( args_info.bank_given )
    {
      cerr << "bank " << args_info.bank_arg << endl;  
      strncpy(account.bank_id,args_info.bank_arg,OFX_BANKID_LENGTH-1);
    }
    else    
    {
      if ( args_info.type_given && args_info.type_arg == 1 )
      {
        cerr << "ERROR: --bank is required for a bank request" << endl;
        ok = false;
      }
    }
    
    if ( args_info.broker_given )
    {
      cerr << "broker " << args_info.broker_arg << endl;  
      strncpy(account.broker_id,args_info.broker_arg,OFX_BROKERID_LENGTH-1);
    }
    else
    {
      if ( args_info.type_given && args_info.type_arg == 2 )
      {
        cerr << "ERROR: --broker is required for an investment statement request" << endl;
        ok = false;
      }
    }
    
    if ( args_info.acct_given )
    {
      cerr << "acct " << args_info.acct_arg << endl;  
      strncpy(account.account_number,args_info.acct_arg,OFX_ACCTID_LENGTH-1);
    }
    else
    {
      cerr << "ERROR: --acct is required for a statement request" << endl;
      ok = false;
    }
    
    if ( args_info.type_given )
    { 
      cerr << "type " << args_info.type_arg << endl;
      switch (args_info.type_arg) {
      case 1: account.account_type = account.OFX_CHECKING;
	break;
      case 2: account.account_type = account.OFX_INVESTMENT;
	break;
      case 3: account.account_type = account.OFX_CREDITCARD ;
	break;
      default:
	cerr << "ERROR: --type is not valid.  Must be between 1 and 3" << endl;
	ok = false;
      }
    }
    else
    {
      cerr << "ERROR: --type is required for a statement request" << endl;
      ok = false;
    }
    
    if ( ok )
    {
      char* request = libofx_request_payment( &fi, &account, &payee, &payment );
    
      filebuf fb;
      fb.open ("query",ios::out);
      ostream os(&fb);
      os << request;
      fb.close();
      
      if ( url.length() ) 
        post(request,url.c_str(),args_info.inputs[0]);
      else
        cout << request;
    
      free(request);
    }
        
  }
  
  if ( args_info.accountinfo_req_given )
  {
    if ( ok )
    {
      char* request = libofx_request_accountinfo( &fi );
    
      if ( url.length() ) 
        post(request,url.c_str(),args_info.inputs[0]);
      else
        cout << request;
    
      free(request);
    }
  }
        
  if ( args_info.bank_list_given )
  {
    cout << OfxPartner::BankNames();
  }
  
  if ( args_info.bank_fipid_given )
  {
    cout << OfxPartner::FipidForBank(args_info.inputs[0]);
  }
  
  if ( args_info.bank_services_given )
  {
    OfxFiServiceInfo svcinfo = OfxPartner::ServiceInfo(args_info.inputs[0]);
    cout << "Account List? " << (svcinfo.accountlist?"Yes":"No") << endl;
    cout << "Statements? " << (svcinfo.statements?"Yes":"No") << endl;
    cout << "Billpay? " << (svcinfo.billpay?"Yes":"No") << endl;
    cout << "Investments? " << (svcinfo.investments?"Yes":"No") << endl;
  }
 
  if ( args_info.allsupport_given )
  {
    vector<string> banks = OfxPartner::BankNames();
    vector<string>::const_iterator it_bank = banks.begin();
    while ( it_bank != banks.end() )
    {
      vector<string> fipids = OfxPartner::FipidForBank(*it_bank);
      vector<string>::const_iterator it_fipid = fipids.begin();
      while ( it_fipid != fipids.end() )
      {
        if ( OfxPartner::ServiceInfo(*it_fipid).accountlist )
          cout << *it_bank << endl;
        ++it_fipid;
      }
      ++it_bank;
    }
  }
  
  return 0;
}


// vim:cin:si:ai:et:ts=2:sw=2:

